solana/write-account/src/program.rs

```diff
@@ -136,63 +110,139 @@ fn handle_write(
+                &[accounts.payer.clone(), accounts.write.clone()],
             )?;
         }
-        write_account.realloc(size, false)
+        accounts.write.realloc(size, false)
     } else {
         // Otherwise, the account exists and is large enough.  There’s nothing
         // we need to do.
         Ok(())
     }
 }
 
+
+/// Handles Free operation.
+fn handle_free(accounts: Accounts) -> Result {
+    {
+        let mut payer = accounts.payer.try_borrow_mut_lamports()?;
+        let mut write = accounts.write.try_borrow_mut_lamports()?;
+        let lamports = payer
+            .checked_add(**write)
+            .ok_or(ProgramError::ArithmeticOverflow)?;
+        **payer = lamports;
+        **write = 0;
+    }
+
+    accounts.write.assign(&solana_program::system_program::ID);
+    accounts.write.realloc(0, false)
+}
+
+
+/// Accounts used when processing instruction.
+#[derive(Clone, Copy)]
+struct Accounts<'a, 'info> {
+    /// The Payer account which pays and ‘owns’ the Write account.
+    payer: &'a AccountInfo<'info>,
+
+    /// The Write account.  It’s address is a PDA using `[payer.key,
+    /// seed_and_bump]` seeds.
+    write: &'a AccountInfo<'info>,
+
+    /// Seed and bump used in PDA of the Write account.
+    seed_and_bump: &'a [u8],
+}
+
+impl<'a, 'info> Accounts<'a, 'info> {
+    /// Gets and verifies accounts for handling the instruction.
+    ///
+    /// Expects the following accounts in the `accounts` slice:
+    /// 1. Payer account which is signer and writable,
+    /// 2. Write account which is writable and a PDA using `[payer.key, seed,
+    ///    bump]` seeds.
+    ///
+    /// Reads seed and bump from `instruction` advancing it.  Specifically,
+    /// reads the following dynamically-sized structure:
+    ///
+    /// ```ignore
+    /// #[repr(C, packed)]
+    /// struct SeedAndBump {
+    ///     seed_len: u8,
+    ///     seed: [u8; seed_len],
+    ///     bump: u8,
+    /// }
+    /// ```
+    fn get(
+        program_id: &Pubkey,
+        accounts: &'a [AccountInfo<'info>],
+        instruction: &mut &'a [u8],
+    ) -> Result<Self> {
+        let accounts = &mut accounts.iter();
+
+        // Payer.  Must be signer and writable.
+        let payer = next_account_info(accounts)?;
+        if !payer.is_signer {
+            return Err(ProgramError::MissingRequiredSignature);
+        } else if !payer.is_writable {
+            return Err(ProgramError::InvalidAccountData);
+        }
+
+        // Write account.  Must be writable and PDA.
+        let write = next_account_info(accounts)?;
+        if !write.is_writable {
+            return Err(ProgramError::InvalidAccountData);
+        }
+        let seed_len = read(instruction, u8::from_le_bytes)?;
+        let seed_and_bump = read_slice(instruction, seed_len as usize + 1)?;
+        let this = Self { payer, write, seed_and_bump };
+
+        match Pubkey::create_program_address(&this.write_seeds(), program_id) {
```

Comment:
Since we are verifying the write program seeds here, we should skip this check again in `setup_write_program` method right? I feel we should also verify bumps here.
https://github.com/ComposableFi/emulated-light-client/blob/f7f64888b41b53f70ed068be96a944577e50ad11/solana/write-account/src/program.rs#L149-L152

Reply:
The check in setup_write_account has indeed been removed.

Reply:
oh sorry, my bad.